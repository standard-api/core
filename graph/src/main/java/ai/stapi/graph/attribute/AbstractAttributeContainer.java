package ai.stapi.graph.attribute;

import ai.stapi.graph.AttributeContainer;
import ai.stapi.graph.exceptions.GraphException;
import ai.stapi.graph.versionedAttributes.ImmutableVersionedAttributeGroup;
import ai.stapi.graph.versionedAttributes.VersionedAttribute;
import ai.stapi.graph.versionedAttributes.VersionedAttributeGroup;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public abstract class AbstractAttributeContainer implements AttributeContainer {

  private final VersionedAttributeGroup versionedAttributes;

  protected AbstractAttributeContainer() {
    this.versionedAttributes = new ImmutableVersionedAttributeGroup();
  }

  protected AbstractAttributeContainer(VersionedAttributeGroup versionedAttributes) {
    this.versionedAttributes = versionedAttributes;
  }

  protected AbstractAttributeContainer(Attribute<?>... attributes) {
    this(new ImmutableVersionedAttributeGroup(attributes));
  }

  @Override
  public <T extends Attribute<?>> AttributeContainer add(T attribute) {
    var newAttributes = this.versionedAttributes.add(attribute);
    return this.withNewAttributes(newAttributes);
  }

  protected abstract AttributeContainer withNewAttributes(VersionedAttributeGroup newAttributes);

  public VersionedAttributeGroup getVersionedAttributes() {
    return versionedAttributes;
  }


  public List<VersionedAttribute<?>> getVersionedAttributeList() {
    return versionedAttributes.getVersionedAttributeList();
  }

  public List<Attribute<?>> getFlattenAttributes() {
    return this.versionedAttributes.getVersionedAttributeList().stream()
        .flatMap(VersionedAttribute::streamAttributeVersions)
        .collect(Collectors.toList());
  }

  @Override
  public VersionedAttribute<?> getVersionedAttribute(String name) {
    return versionedAttributes.getVersionedAttribute(name);
  }

  @Override
  public Attribute<?> getAttribute(String name) throws GraphException {
    return this.versionedAttributes.getCurrentAttribute(name);
  }

  @Override
  public boolean containsAttribute(String name, Object value) {
    return this.versionedAttributes.isValuePresentInAnyVersion(name, value);
  }

  @Override
  public boolean containsAttribute(String name) {
    return this.versionedAttributes.hasAttribute(name);
  }

  @Override
  public Object getAttributeValue(String name) {
    return this.versionedAttributes.getCurrentAttributeValue(name);
  }

  public AttributeContainer mergeAttributesWithAttributesOf(AbstractAttributeContainer other) {
    return this.withNewAttributes(
        this.versionedAttributes.mergeOverwrite(other.getVersionedAttributes()));
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof AbstractAttributeContainer that)) {
      return false;
    }
    return getVersionedAttributes().equals(that.getVersionedAttributes());
  }

  @Override
  public int hashCode() {
    return Objects.hash(this.getVersionedAttributes()) + this.getHashCodeWithoutAttributes();
  }

  public int getIdlessHashCode() {
    VersionedAttributeGroup versionedAttributes1 = this.getVersionedAttributes();
    int idlessHashCodeWithoutAttributes = this.getIdlessHashCodeWithoutAttributes();
    return Objects.hash(versionedAttributes1) + idlessHashCodeWithoutAttributes;
  }

  protected abstract int getHashCodeWithoutAttributes();

  protected abstract int getIdlessHashCodeWithoutAttributes();

  @Override
  public boolean hasAttribute(String name) {
    return this.versionedAttributes.hasAttribute(name);
  }

  public int getAttributeCount() {
    return this.versionedAttributes.numberOfUniqueAttributeNames();
  }


  public String guessBestName() {
    if (this.hasAttribute("name")) {
      return (String) this.getAttribute("name").getValue();
    }
    return null;
  }
}
