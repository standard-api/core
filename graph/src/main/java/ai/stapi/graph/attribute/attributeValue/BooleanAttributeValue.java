package ai.stapi.graph.attribute.attributeValue;

import org.jetbrains.annotations.NotNull;

public class BooleanAttributeValue extends AbstractAttributeValue<Boolean> {

  public static final String SERIALIZATION_TYPE = "boolean";

  public BooleanAttributeValue(Boolean value) {
    super(value);
  }

  @Override
  public String getDataType() {
    return SERIALIZATION_TYPE;
  }

  @Override
  public int compareTo(@NotNull AttributeValue<?> other) {
    if (other instanceof BooleanAttributeValue otherBooleanValue) {
      return this.getValue().compareTo(otherBooleanValue.getValue());
    } else {
      return -1;
    }
  }
}
