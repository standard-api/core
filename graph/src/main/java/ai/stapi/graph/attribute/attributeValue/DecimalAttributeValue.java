package ai.stapi.graph.attribute.attributeValue;

public class DecimalAttributeValue extends AbstractAttributeValue<Double> implements NumberLikeAttributeValue<Double> {

  public static final String SERIALIZATION_TYPE = "decimal";

  public DecimalAttributeValue(Double value) {
    super(value);
  }

  @Override
  public String getDataType() {
    return SERIALIZATION_TYPE;
  }

  @Override
  public Double toDoubleValue() {
    return this.getValue();
  }
}
