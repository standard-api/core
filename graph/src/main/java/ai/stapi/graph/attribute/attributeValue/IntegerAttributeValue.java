package ai.stapi.graph.attribute.attributeValue;

public class IntegerAttributeValue extends AbstractAttributeValue<Integer> implements NumberLikeAttributeValue<Integer> {

  public static final String SERIALIZATION_TYPE = "integer";

  public IntegerAttributeValue(Integer value) {
    super(value);
  }

  @Override
  public String getDataType() {
    return SERIALIZATION_TYPE;
  }

  @Override
  public Double toDoubleValue() {
    return this.getValue().doubleValue();
  }
}
