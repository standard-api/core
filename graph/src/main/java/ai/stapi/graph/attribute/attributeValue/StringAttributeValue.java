package ai.stapi.graph.attribute.attributeValue;

public class StringAttributeValue extends AbstractAttributeValue<String> implements StringLikeAttributeValue<String> {

  public static final String SERIALIZATION_TYPE = "string";

  public StringAttributeValue(String value) {
    super(value.strip());
  }

  @Override
  public String getDataType() {
    return SERIALIZATION_TYPE;
  }

  @Override
  public String toStringValue() {
    return this.getValue();
  }
}
