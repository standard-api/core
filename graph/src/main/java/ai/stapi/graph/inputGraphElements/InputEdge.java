package ai.stapi.graph.inputGraphElements;

import ai.stapi.graph.AttributeContainer;
import ai.stapi.graph.NodeIdAndType;
import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.inMemoryGraph.exceptions.GraphEdgesCannotBeMerged;
import ai.stapi.graph.traversableGraphElements.TraversableEdge;
import ai.stapi.graph.versionedAttributes.VersionedAttributeGroup;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import java.util.Objects;

public class InputEdge extends AbstractGraphElement {

  private final UniqueIdentifier nodeFromId;
  private final UniqueIdentifier nodeToId;
  private final String nodeFromType;
  private final String nodeToType;

  public InputEdge(
      UniqueIdentifier id,
      String type,
      VersionedAttributeGroup versionedAttributes,
      NodeIdAndType nodeFromIdAndType,
      NodeIdAndType nodeToIdAndType
  ) {
    super(id, type, versionedAttributes);
    this.nodeFromId = nodeFromIdAndType.getId();
    this.nodeToId = nodeToIdAndType.getId();
    this.nodeFromType = nodeFromIdAndType.getType();
    this.nodeToType = nodeToIdAndType.getType();
  }

  public InputEdge(
      UniqueIdentifier edgeId,
      InputNode inputNodeFrom,
      String edgeType,
      InputNode inputNodeTo
  ) {
    super(edgeId, edgeType);
    this.nodeFromId = inputNodeFrom.getId();
    this.nodeFromType = inputNodeFrom.getType();
    this.nodeToId = inputNodeTo.getId();
    this.nodeToType = inputNodeTo.getType();
  }

  public InputEdge(
      UniqueIdentifier edgeId,
      String edgeType,
      InputNode inputNodeFrom,
      InputNode inputNodeTo,
      VersionedAttributeGroup attributeGroup
  ) {
    super(edgeId, edgeType, attributeGroup);
    this.nodeFromId = inputNodeFrom.getId();
    this.nodeFromType = inputNodeFrom.getType();
    this.nodeToId = inputNodeTo.getId();
    this.nodeToType = inputNodeTo.getType();
  }

  public InputEdge(
      UniqueIdentifier edgeId,
      InputNode inputNodeFrom,
      String edgeType,
      InputNode inputNodeTo,
      Attribute<?>... attributes
  ) {
    super(edgeId, edgeType, attributes);
    this.nodeFromId = inputNodeFrom.getId();
    this.nodeFromType = inputNodeFrom.getType();
    this.nodeToId = inputNodeTo.getId();
    this.nodeToType = inputNodeTo.getType();
  }

  public InputEdge(TraversableEdge traversableEdge) {
    this(
        traversableEdge.getId(),
        traversableEdge.getType(),
        new InputNode(traversableEdge.getNodeFrom()),
        new InputNode(traversableEdge.getNodeTo()),
        traversableEdge.getVersionedAttributes()
    );
  }

  public InputEdge(InputNode inputNodeFrom, String edgeType, InputNode inputNodeTo) {
    this(
        UniversallyUniqueIdentifier.randomUUID(),
        inputNodeFrom,
        edgeType,
        inputNodeTo
    );
  }

  public InputEdge(
      InputNode inputNodeFrom,
      String edgeType,
      InputNode inputNodeTo,
      VersionedAttributeGroup attributeGroup
  ) {
    this(UniversallyUniqueIdentifier.randomUUID(), edgeType, inputNodeFrom, inputNodeTo,
        attributeGroup);
  }

  public InputEdge(
      InputNode inputNodeFrom,
      String edgeType,
      InputNode inputNodeTo,
      Attribute<?>... attributes
  ) {
    this(UniversallyUniqueIdentifier.randomUUID(), inputNodeFrom, edgeType, inputNodeTo,
        attributes);
  }

  public InputEdge(
      UniqueIdentifier edgeId,
      String edgeType,
      UniqueIdentifier nodeFromId,
      UniqueIdentifier nodeToId,
      String nodeFromType,
      String nodeToType,
      VersionedAttributeGroup versionedAttributes
  ) {
    super(edgeId, edgeType, versionedAttributes);
    this.nodeFromId = nodeFromId;
    this.nodeToId = nodeToId;
    this.nodeFromType = nodeFromType;
    this.nodeToType = nodeToType;
  }

  public UniqueIdentifier getNodeFromId() {
    return this.nodeFromId;
  }

  public UniqueIdentifier getNodeToId() {
    return this.nodeToId;
  }

  public String getNodeFromType() {
    return this.nodeFromType;
  }

  public String getNodeToType() {
    return this.nodeToType;
  }

  @Override
  protected AttributeContainer withNewAttributes(VersionedAttributeGroup newAttributes) {
    return new InputEdge(
        this.getId(),
        this.getType(),
        this.getNodeFromId(),
        this.getNodeToId(),
        this.getNodeFromType(),
        this.getNodeToType(),
        newAttributes
    );
  }

  public <T extends Attribute<?>> InputEdge addToEdge(T attribute) {
    return (InputEdge) this.add(attribute);
  }

  @Override
  protected int getHashCodeWithoutAttributes() {
    return Objects.hash(this.getId()) + this.getIdlessHashCodeWithoutAttributes();
  }

  @Override
  protected int getIdlessHashCodeWithoutAttributes() {
    return Objects.hash(this.getType());
  }

  public NodeIdAndType getNodeFromIdAndType() {
    return new NodeIdAndType(this.nodeFromId, this.nodeFromType);
  }

  public NodeIdAndType getNodeToIdAndType() {
    return new NodeIdAndType(this.nodeToId, this.nodeToType);
  }

  public InputEdge getCopy() {
    return this;
  }

  public InputEdge mergeOverwrite(InputEdge otherEdge) {
    if (!this.getId().equals(otherEdge.getId())) {
      throw GraphEdgesCannotBeMerged.becauseTheyHaveDifferentIds();
    }
    if (!this.getType().equals(otherEdge.getType())) {
      throw GraphEdgesCannotBeMerged.becauseTheyHaveDifferentTypes();
    }
    if (!(this.getNodeFromId().equals(otherEdge.getNodeFromId()))
        || !(this.getNodeToId().equals(otherEdge.getNodeToId()))) {
      throw GraphEdgesCannotBeMerged.becauseTheyHaveDifferentNodeIds();
    }
    return (InputEdge) this.mergeAttributesWithAttributesOf(otherEdge);
  }
}
