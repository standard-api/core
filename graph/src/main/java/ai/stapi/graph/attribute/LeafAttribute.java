package ai.stapi.graph.attribute;


import ai.stapi.graph.attribute.attributeValue.AttributeValue;
import java.sql.Timestamp;
import java.util.Map;

public class LeafAttribute<T, AV extends AttributeValue<T>> extends AbstractAttribute<T> {

  public static final String DATA_STRUCTURE_TYPE = "LeafAttribute";
  private final AV boxedValue;

  public LeafAttribute(String name, Map<String, MetaData> metaData, AV boxedValue) {
    super(name, metaData);
    this.boxedValue = boxedValue;
  }

  public LeafAttribute(String name, Timestamp createdAt, AV boxedValue) {
    super(name, createdAt);
    this.boxedValue = boxedValue;
  }

  public LeafAttribute(String name, AV boxedValue) {
    super(name);
    this.boxedValue = boxedValue;
  }

  public AV getBoxedValue() {
    return boxedValue;
  }

  @Override
  public T getValue() {
    return this.getBoxedValue().getValue();
  }

  @Override
  public LeafAttribute<T, AV> getCopy() {
    return new LeafAttribute<>(this.getName(), this.getCreatedAt(), this.getBoxedValue());
  }
}
