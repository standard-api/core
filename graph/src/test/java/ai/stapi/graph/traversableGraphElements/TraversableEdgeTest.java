package ai.stapi.graph.traversableGraphElements;

import ai.stapi.graph.Graph;
import ai.stapi.graph.NullNodeLoader;
import ai.stapi.graph.attribute.AbstractAttributeContainer;
import ai.stapi.graph.attribute.AttributeContainerTest;
import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.BooleanAttributeValue;
import ai.stapi.graph.attribute.attributeValue.DecimalAttributeValue;
import ai.stapi.graph.attribute.attributeValue.IntegerAttributeValue;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graph.inputGraphElements.InputEdge;
import ai.stapi.graph.inputGraphElements.InputNode;
import ai.stapi.graph.versionedAttributes.ImmutableVersionedAttributeGroup;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

class TraversableEdgeTest extends AttributeContainerTest {

  @Override
  protected AbstractAttributeContainer getAttributeContainer() {
    return new TraversableEdge(
        new InputNode("test_node_from"),
        "edge_type",
        new InputNode("test_node_to")
    );
  }

  @Test
  public void itCanCreateEdge_WithInputNodes() {
    //Given
    var expectedEdgeType = "edge_type";
    //When
    var inputNodeFrom = new InputNode("input_node_from");
    var inputNodeTo = new InputNode("input_node_to");
    var edge = new TraversableEdge(
        inputNodeFrom,
        expectedEdgeType,
        inputNodeTo
    );
    //Then
    Assertions.assertEquals(expectedEdgeType, edge.getType());
    Assertions.assertEquals(inputNodeFrom.getId(), edge.getNodeFromId());
    Assertions.assertEquals(inputNodeTo.getId(), edge.getNodeToId());
    Assertions.assertEquals(inputNodeFrom.getType(), edge.getNodeFromType());
    Assertions.assertEquals(inputNodeTo.getType(), edge.getNodeToType());
  }

  @Test
  public void itCanCreateEdge_WithVersionedAttributes() {
    //Given
    var expectedEdgeType = "edge_type";
    var expectedNodeFrom = new TraversableNode("irrelevant");
    var expectedNodeTo = new TraversableNode("irrelevant");
    var expectedVersionedAttributeGroup = new ImmutableVersionedAttributeGroup(
        new LeafAttribute<>("test_boolean", new BooleanAttributeValue(true)),
        new LeafAttribute<>("test_double", new DecimalAttributeValue(10d)),
        new LeafAttribute<>("test_integer", new IntegerAttributeValue(15))
    );
    //When
    var node = new TraversableEdge(
        UniversallyUniqueIdentifier.randomUUID(),
        expectedNodeFrom,
        expectedEdgeType,
        expectedNodeTo,
        expectedVersionedAttributeGroup,
        new NullNodeLoader()
    );
    //Then
    Assertions.assertEquals(
        expectedEdgeType,
        node.getType()
    );
    Assertions.assertEquals(
        expectedNodeFrom.getId(),
        node.getNodeFromId()
    );
    Assertions.assertEquals(
        expectedNodeFrom.getType(),
        node.getNodeFromType()
    );
    Assertions.assertEquals(
        expectedNodeTo.getId(),
        node.getNodeToId()
    );
    Assertions.assertEquals(
        expectedNodeTo.getType(),
        node.getNodeToType()
    );
    Assertions.assertEquals(
        expectedVersionedAttributeGroup,
        node.getVersionedAttributes()
    );
  }

  @Test
  public void itHasDeterministicHashCode() {

    var versionedAttributeGroup = new ImmutableVersionedAttributeGroup(
        new LeafAttribute<>("test_boolean", new BooleanAttributeValue(true)),
        new LeafAttribute<>("test_double", new DecimalAttributeValue(10d)),
        new LeafAttribute<>("test_integer", new IntegerAttributeValue(15))
    );

    var node1 = new InputNode("irrelevant");
    var node2 = new InputNode("irrelevant");
    var savedEdge = new InputEdge(
        UniversallyUniqueIdentifier.fromString("0bac1abc-660e-4e92-94e2-3e8953873c58"),
        "edge_type",
        node1,
        node2,
        versionedAttributeGroup
    );

    var graph = new Graph(
        node1,
        node2,
        savedEdge
    );

    var loadedEdge = graph.getEdge(
        savedEdge.getId(),
        savedEdge.getType()
    );

    var firstActualHashCode = loadedEdge.getIdlessHashCode();
    var secondActualHashCode = loadedEdge.getIdlessHashCode();

    Assertions.assertEquals(
        firstActualHashCode,
        secondActualHashCode
    );
    Assertions.assertEquals(
        loadedEdge.hashCode(),
        loadedEdge.hashCode()
    );
    Assertions.assertEquals(
        1718053182,
        firstActualHashCode
    );
  }

  @Test
  public void itCanGetNodeFromFromBuilder() {
    //Given
    var expectedEdgeType = "edge_type";
    var expectedNodeFrom = new InputNode(
        "from_node_type",
        new LeafAttribute<>("test_string", new StringAttributeValue("NodeFrom attribute value"))
    );
    var expectedNodeTo = new InputNode(
        "to_node_type",
        new LeafAttribute<>("test_string", new StringAttributeValue("NodeTo attribute value"))
    );
    //When
    var edge = new InputEdge(
        expectedNodeFrom,
        expectedEdgeType,
        expectedNodeTo
    );

    var graph = new Graph(
        expectedNodeFrom,
        expectedNodeTo,
        edge
    );

    var traversableEdge = graph.traversable().loadEdge(
        edge.getId(),
        edge.getType()
    );

    TraversableNode nodeFrom = traversableEdge.getNodeFrom();
    this.thenNodeApproved(nodeFrom);
  }

  @Test
  public void itCanGetNodeToFromBuilder() {
    //Given
    var expectedEdgeType = "edge_type";
    var expectedNodeFrom = new InputNode(
        "from_node_type",
        new LeafAttribute<>("test_string", new StringAttributeValue("NodeFrom attribute value"
        ))
    );
    var expectedNodeTo = new InputNode(
        "to_node_type",
        new LeafAttribute<>("test_string", new StringAttributeValue("NodeTo attribute value"
        ))
    );
    //When
    var edge = new InputEdge(
        expectedNodeFrom,
        expectedEdgeType,
        expectedNodeTo
    );

    var graph = new Graph(
        expectedNodeFrom,
        expectedNodeTo,
        edge
    );

    var traversableEdge = graph.traversable().loadEdge(
        edge.getId(),
        edge.getType()
    );

    this.thenNodeApproved(traversableEdge.getNodeTo());
  }
}
