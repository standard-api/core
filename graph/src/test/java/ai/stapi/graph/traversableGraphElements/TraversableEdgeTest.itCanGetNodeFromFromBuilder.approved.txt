node_type: from_node_type
node_hash: C094C
node_edges:
	from_node_type -> edge_type -> to_node_type (C094C -> FD142 -> 70F19)
attributes:
	test_string -> NodeFrom attribute value
